### ISSUE TITLE
LogArray::nums_for_index questions

### ISSUE THREAD

**Issue by spl** [terminusdb/terminus-store#34]
I'm trying to understand the `nums_for_index` function, in particular this part:

https://github.com/terminusdb/terminus-store/blob/7dfb9382d725936a17c5eff179441723cfad885d/src/structure/logarray.rs#L106-L115

Is it possible for `start_u64_offset + 16 > self.len_bytes` in legal data? When would this happen? Why does it make sense to call `BigEndian::read_u64` on data in which some number of the bytes are zeroed by `vec![0; 16]`?

**Comment by matko** [comment]
Elements in a logarray have a bit width which doesn't have to divide 64 bit cleanly. For example, it could be storing 5-bit elements. This means that some elements may cross 64 bit boundaries. Considering the logarray as a sequence of 64 bit numbers, some elements will be spread over two subsequent 64 bit numbers. `nums_for_index` will retrieve two adjacent 64 bit numbers, ensuring that other code always has the same kind of data to work with.

At the end of a logarray though, we may run out of data. `nums_for_index` wants there to be 16 bytes, but there may be less. So the code is there to ensure that there's always 16 bytes to work with.

`len_bytes` stores the amount of bytes that are actually used to store all the elements. It is definitely possible for `start_u64_offset+16` to be greater than `self.len_bytes`once you're near the end of the logarray. That said, `LogArrayFileBuilder::finalize` will always pad to 8 bytes. This means that there should actually always be enough data to read the first number, and the second number should always be 0.

It's been a while since I looked at this code. I suspect I wrote `nums_for_index` defensively, assuming a non-padded byte array, but then ended up padding anyway. It can probably be simplified.
